### DISCUSSION TITLE
Lift `NoCell` requirements on exclusive references

### DISCUSSION THREAD

**Discussion by jswrenn** [google/zerocopy#1130] [Ideas]
We believe it may be sound to eliminate our `NoCell` requirements on methods that consume exclusive references, such as `FromBytes::mut_from` and `transmute_mut!`. Our reasoning for this is that `NoCell` guards against the risk of data races (two references to the same memory cannot be concurrently mutated with different synchronization strategies), but exclusivity *also* prevents data races (because there is only one active reference to that memory).

We're looking for assurance that this reasoning is correct in https://github.com/rust-lang/unsafe-code-guidelines/issues/495. Once we have that, we can proceed with implementation.

In the simplest case, `transmute_mut`, removing the `NoCell` bound is as simple as *removing the `NoCell` bound* (#1049). However, for our APIs that build atop our internal abstractions, removing `NoCell` bounds is trickier. Our approach here falls into two categories:
1. modify documented safety preconditions (e.g., #999)
2. encode our datarace-related reasoning into the type system

We call this data-race reasoning "Cell Safety", and encode this reasoning into the type system with a trait called `CellSafe`. The `CellSafe` trait encodes the justification for why a combination of pointer invariants means a data race cannot occur: https://github.com/google/zerocopy/commit/14212505bf9b30bb80ac114cad5f7112d8d681f0#diff-c84b1c1033a8a168bf891d71b31d02bad91617845f66de445f96c820a6b2bfb9

This technique allows us to avoid providing two sets of methods, each with a different cell safety reason; i.e., one that consumes an exclusive pointer, and another that consumes a shared pointer but has a `NoCell` bound. In #1026 we apply this approach to parameterizing `TryFromBytes::is_bit_valid`.

**Comment by joshlf** [comment]
This was implemented #1211, #1224, #1225, and #1236.
